### PR TITLE
Make matching by header work

### PR DESCRIPTION
**Pull request by pweinger**
Since extra headers are ignored, a request with 0 and 1 headers both match and would be returned.
This prioritizes matching more headers
Makes sure that this works in both standard and validation modes

### CODE REVIEW

File: src/lib/parse/contracts.js [hunk 1]
```diff
@@ -243,12 +243,16 @@ type BodyMetadata = {
...
     try {
-        // the validator throws an error with empty contract  
-        // this validates when you have an empty body in this case
-        if (!bodyDescriptor.body && !schema) {
-            result = { valid: true, niceErrors: [] };
+        // the validator throws an error with empty scenario body  
+        // this validates this case
+        if (!bodyDescriptor.body) {
+            if (schema) {
+                result = { valid: false, niceErrors: ["No response body found"] };
```
Comment (pweinger) [inline src/lib/parse/contracts.js]: This is just improving the error message to be more user friendly

File: src/lib/handler-filter.js [hunk 2]
```diff
@@ -30,8 +38,12 @@ exports.filterHandlers = function(req, handlers, ignoreHeaders) {
         }
 
         filteredHandlers = handlers.filter(filterRequestHeader(req, ignoreHeaders));
+        
+        // prioritize handlers that have more headers where all headers match. 
+        // this allows us to safely ignore extra headers being sent
+        filteredHandlers.sort((handler1, handler2) => requestHeaderCount(handler2) - requestHeaderCount(handler1));
```
Comment (halfvector) [inline src/lib/handler-filter.js]: Thanks for explaining this

File: src/lib/parse/contracts.js (hunk 1 above)
Comment (halfvector) [reply src/lib/parse/contracts.js]: 👍

File: src/lib/parse/contracts.js [hunk 3]
```diff
@@ -279,10 +283,13 @@ const removeInvalidFixtures = (resource: BlueprintResource, contractActions: ?Ac
         action.examples.forEach((example, exampleIndex) => {
             const validRequests: Array<BodyDescriptor> = [];
             const validResponses: Array<BodyDescriptor> = [];
+
+            //expect requests and responses in pairs - otherwise it is hard to reason about author's expectations
+            if (example.requests.length > 1 ||  example.responses.length > 1) {
+                throw new Error(`Found more than one request or response for example ${exampleIndex}. Requests and responses expected in pairs.`);
+            }
```
Comment (halfvector) [inline src/lib/parse/contracts.js]: Yes, reducing surprises!

File: src/test/api/contract-scenario-headers-test.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            describe('WHEN calling without a header', () => {
+                it('THEN it returns the correct response ', (done) => {
+                    request.get('/headers')
+                        .expect(200)
+                        .expect('"No Header"\n')
+                        .end(done);
+                });
+            });
+
+            describe('WHEN calling without a header', () => {
```
Comment (halfvector) [inline src/test/api/contract-scenario-headers-test.js]:
```suggestion
            describe('WHEN calling with a header', () => {
```

**Pull request merged**